src/java.base/share/native/libjava/io_util.c

```diff
@@ -91,28 +94,44 @@ readBytes(JNIEnv *env, jobject this, jbyteArray bytes,
 
     if (len == 0) {
         return 0;
-    } else if (len > BUF_SIZE) {
-        buf = malloc(len);
+    } else if (len > STACK_BUF_SIZE) {
+        buf_size = len < MAX_MALLOC_SIZE ? len : MAX_MALLOC_SIZE;
+        buf = malloc(buf_size);
         if (buf == NULL) {
             JNU_ThrowOutOfMemoryError(env, NULL);
             return 0;
         }
     } else {
         buf = stackBuf;
+        buf_size = STACK_BUF_SIZE;
     }
 
-    fd = getFD(env, this, fid);
-    if (fd == -1) {
-        JNU_ThrowIOException(env, "Stream Closed");
-        nread = -1;
-    } else {
-        nread = IO_Read(fd, buf, len);
-        if (nread > 0) {
-            (*env)->SetByteArrayRegion(env, bytes, off, nread, (jbyte *)buf);
-        } else if (nread == -1) {
-            JNU_ThrowIOExceptionWithLastError(env, "Read error");
-        } else { /* EOF */
+    nread = 0;
+    while (nread < len) {
+        read_size = len - nread;
+        if (read_size > buf_size)
+            read_size = buf_size;
+        fd = getFD(env, this, fid);
+        if (fd == -1) {
+            JNU_ThrowIOException(env, "Stream Closed");
             nread = -1;
+            break;
+        }
+        n = IO_Read(fd, buf, read_size);
+        if (n > 0) {
+            (*env)->SetByteArrayRegion(env, bytes, off, n, (jbyte*)buf);
+            nread += n;
+            // Exit loop on short read
+            if (n < read_size)
+                break;
+            off += n;
+        } else if (n == -1) {
+            JNU_ThrowIOExceptionWithLastError(env, "Read error");
```

Comment:
The original code would have `nread` set to `-1` here, and thus exit with `-1`.  The new code would exit with the last value for `nread` which could be anything.